packages/framework/presence/src/systemWorkspace.ts

```diff
@@ -149,22 +150,23 @@ class SystemWorkspaceImpl implements PresenceStatesInternal, SystemWorkspace {
 				/* order */ value.rev,
 			);
 
-			// Check new attendee against audience to see if they're currently connected
-			const isAttendeeConnected = audienceMembers.has(clientConnectionId);
+			// Check new attendee to see if they're currently connected
+			const isAttendeeConnected =
+				audienceMembers.has(clientConnectionId) ||
+				senderConnectionId === clientConnectionId ||
+				connectedAttendees.has(attendee);
 
 			if (isAttendeeConnected) {
+				// If attendee is connected, update their connection ID and status.
 				connectedAttendees.add(attendee);
-				if (attendee.getConnectionStatus() === SessionClientStatus.Disconnected) {
-					attendee.setConnectionId(clientConnectionId);
-				}
-				if (isNew) {
-					// If the attendee is both new and in audience (i.e. currently connected), emit an attendeeJoined event.
+				attendee.setConnectionId(clientConnectionId);
+				if (isNew && !announcedAttendees.has(clientSessionId)) {
+					// If the attendee is both new and currently connected, emit an attendeeJoined event.
 					postUpdateActions.push(() => this.events.emit("attendeeJoined", attendee));
+					announcedAttendees.add(clientSessionId);
 				}
```

Comment:
You could skip the !has check if you just added to set and then processed the set afterward to generate announcements.

Reply:
Updated